content/courses/intro-to-embedded-sys-prog/chapters/low_level_programming.rst

```diff
@@ -579,41 +657,919 @@ sending values over the wire, the code should query the size of the
+      for List'Size use 24 * 3;
+   
+   end Some_Types;
+     
+When we compile this, the compiler will complain that the size for 
+"List" is too small, i.e., that the minimum allowed is 96 bits instead 
+of the 72 we specified. We specified 24 * 3 because we said the record 
+size should be 24 bits, and we want our array to contain 3 record 
+components of that size, so 72 seems right.
+
+What's wrong? As we've shown earlier, specifying the record type size 
+doesn't necessarily mean that objects (in this case array components) 
+are that size. The object size could be bigger than we specified for the 
+type. In this case, the compiler says we need 96 total bits for the 
+array type, meaning that each of the 3 array components is 32 bits wide 
+instead of 24. 
+
+Why is it 32 bits? Because the alignment for Info is 2 (on this 
+machine). The record alignment is a multiple of the largest 
+alignment of the enclosed components. The alignment for type Temperature 
+(2), is larger than the alignment for type Identity (1), therefore the 
+alignment for the whole record type is 2. We need to go from that number 
+of storage elements to a number of bits for the size. 
+
+Here's where it gets subtle. The alignment is in terms of storage 
+elements. Each storage element is of a size in bits given by 
+System.Storage_Unit. We've said that on our hypothetical machine 
+Storage_Unit is 8, so storage elements are 8 bits wide on this machine. 
+Bytes, in other words. Therefore, to get the required size in bits, we 
+have to find a multiple of the two 8-bit bytes (specified by the 
+alignment) that has at least the number of bits we gave in the Size 
+clause. Two bytes only provides 16 bits, so that's not big enough, we 
+need at least 24 bits. The next multiple of 2 bytes is 4 bytes, 
+providing 32 bits, which is indeed larger than 24. Therefore, the 
+overall size of the record type, consistent with the alignment, is 4 
+bytes, or 32 bits. That's why the compiler says each array component is 
+32 bits wide. 
+
+But for our example let's say that we really want to use only 72 total 
+bits for the array type (and that we want three array components). 
+That's the size we specified, after all. So how do we get the record 
+type to be 24 bits instead of 32? Yes, you guessed it, we change the 
+alignment for the record type. If we change it from 2 to 1, the size of 
+24 bits will work. Adding this Alignment clause line will do that: 
+
+.. code-block:: ada
+
+   for Info'Alignment use 1;
+
+An alignment of 1 means that any address will work, assuming that 
+addresses refer to entire storage elements. (An alignment of 0 would 
+mean that the address need not start on a storage element boundary, but 
+we know of no such machines.) 
+
+We can even entirely replace the Size clause with the Alignment clause, 
+because the Size clause specifying 24 bits is just confirming: it's the 
+value that 'Size would return anyway. The problem is the object size. 
```

Comment:
```suggestion
We can even entirely replace the :ada:`Size` clause with the :ada:`Alignment` clause, 
because the :ada:`Size` clause specifying 24 bits is just confirming: it's the 
value that :ada:`'Size` would return anyway. The problem is the object size. 
```